Catalog revision (#25)
ezsystems/commerce-documentation: docs/guide/catalog/product_variants/setting_up_variants_from_external_source.md

# Setting up variants from external source

The default data provider and `CatalogFactory` ([content model data provider](../../data_providers/content_model_dataprovider.md))

> already support variants up to two levels which can be set up [using a special variant Field Type](product_variants.md) in the Back Office. 

Reviewer: ```suggestion
already support variants up to two levels.
You can set the variants up by [using a variant Field Type](product_variants.md) in the Back Office. 
```

You can also set up variants if you are not using the default Content Types
and don't want to set up variants manually in the Back Office.

To do this, you need to map your variant structure to the [`VariantProductNode`](product_variants.md#variantproductsnode) structure.

!!! tip
    
    The main job of every catalog factory is to fill the [`VariantProductNode`](product_variants.md#variantproductsnode) attribute `variantCharacteristics`.
    The basic concept is to take information from the provided `dataMap`
    (e.g. coming from the data provider, either from the content model or from an external source)
    and create an attribute `variantCharacteristic` for the `VariantProductNode`.

## Mapping variant information

### 1. Create data provider and catalog factory

> The first thing you need to do to create variants is to create a custom data provider and factory for catalog elements. 

Reviewer: ```suggestion
To create variants. the first thing you must do is create a custom data provider and factory for catalog elements. 
```

    It is good practice to configure which method should be used depending on e.g. the class identifier (Content Type).
    This is the default configuration:

    ``` yaml
    silver_eshop.default.catalog_factory.ses_category: createCatalogNode
    silver_eshop.default.catalog_factory.ses_product: createOrderableProductNode
    ```

> `createCatalogElement()` must additionally decide when to call the method `createOrderableProductNode()` and when `createVariantProductNode()`

Reviewer: ```suggestion
`createCatalogElement()` must additionally decide whether to call the `createOrderableProductNode()` or the `createVariantProductNode()` method
```

> because in the example below (standard implementation), the variants are stored as Content items of the same Content Type as regular products (`ses_product`).

Reviewer: "...stored as Content items of the same Content Type as regular products..."
should it be:
"...stored as Content items of the same Content Type, rather than regular products..."
?

> because in the example below (standard implementation), the variants are stored as Content items of the same Content Type as regular products (`ses_product`).

Author: ```suggestion
because in the example below (standard implementation), the variants are stored, like regular products, in Product Content items (`ses_product`).
```

> In the default implementation the factory calls the `createVariantProductNode()` method whenever [eZ Matrix with variants](product_variants.md) is filled.

Reviewer: ```suggestion
In the default implementation, the factory calls the `createVariantProductNode()` method whenever [eZ Matrix with variants](product_variants.md) is filled.
```
